drivers/SPI.cpp

```diff
@@ -16,97 +16,144 @@
 #include "drivers/SPI.h"
 #include "platform/mbed_critical.h"
 
-#if DEVICE_SPI_ASYNCH
+#if 0
 #include "platform/mbed_power_mgmt.h"
 #endif
 
 #if DEVICE_SPI
 
 namespace mbed {
 
-#if DEVICE_SPI_ASYNCH && TRANSACTION_QUEUE_SIZE_SPI
+#if 0 && TRANSACTION_QUEUE_SIZE_SPI
 CircularBuffer<Transaction<SPI>, TRANSACTION_QUEUE_SIZE_SPI> SPI::_transaction_buffer;
 #endif
+SPI::spi_peripheral_s SPI::_peripherals[SPI_COUNT];
 
 SPI::SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel) :
-    _spi(),
-#if DEVICE_SPI_ASYNCH
-    _irq(this),
+    _self(),
+#if 0
     _usage(DMA_USAGE_NEVER),
     _deep_sleep_locked(false),
 #endif
     _bits(8),
-    _mode(0),
+    _mode(SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE),
+    _msb_first(true),
     _hz(1000000),
-    _write_fill(SPI_FILL_CHAR)
+    _write_fill(0xFF)
 {
     // No lock needed in the constructor
+    // get_module
 
-    spi_init(&_spi, mosi, miso, sclk, ssel);
-    _acquire();
+    SPIName name = spi_get_module(mosi, miso, sclk);
+
+    // this is a lazy allocation for the mutex.
+    // unless not assigned to a peripheral, this mutex will never be deleted.
+    PlatformMutex *mutex = new PlatformMutex();
+
+    core_util_critical_section_enter();
+    // lookup in a critical section if we already have it else initialize it
+    _self = SPI::lookup(name, true);
+    if (_self->mutex == NULL) {
+        _self->mutex = mutex;
+    }
+    if (_self->name == 0) {
+        _self->name = name;
+        spi_init(&_self->spi, false, mosi, miso, sclk, NC);
```

Comment:
On further reflection, probably best to not put it in actual `lock`, as we do have a common pattern where `lock` is just a "peripheral lock", and it's virtual so that people can inherit and override, in particular for bypassing the mutex for use from interrupt, if they know they're the only user of the peripheral.

There are some `UnlockedSPI` out there that inherit from `SPI` for use from IRQ, overriding `lock`+`unlock`. Shouldn't break that pattern.

Instead, having `select` and `deselect` methods that are 'lock; acquire; cs=1 (reference counted)' would be the way to go - that would still work when someone overrides lock to eliminate the peripheral mutex.

Reply:
My bad, this NC should be ssel.

Any work regarding locks/selections is out of scope from this PR though.

Reply:
It has to be `NC` - can't be anything else or the entire "peripheral" concept won't work. That inherently prevents any use of the HAL ssel, because you're sharing `spi_t` for the same bus.

Reply:
Well, I guess you /could/ pass ssel to the HAL, but only if you did it on every acquire. But that means a full `init` every time you change owner of a peripheral, I guess?

Reply:
> Any work regarding locks/selections is out of scope from this PR though.

From my fairly prolonged personal experience of using the SPI API, the two biggest problems are the system-wide transfer mutex, and the total uselessness of the SSEL parameter.   (Lack of trust of the async API implementations would be a runner-up)

Great that you're fixing the first one - which is also a purely C++ layer and non-HAL thing - but I don't see how you can do any sort of SPI rework that doesn't address the SSEL problem. If it's not going to be addressed when during an SPI API rework, when is it going to be addressed?

The peripheral change as you've written it here is effectively is going to force some SSEL change. You're going to have to do something. Any of these are possible:

* render SSEL totally non-functional for all platforms, as the current PR, everyone does SSEL manually via GPIO, as portable code already does anyway.
* rework the acquire mechanism to do `spi_init` to change SSEL
* give up the shared `spi_t`
* put the GPIO for SSEL inside `SPI` - doesn't break anyone with non-portable code relying on the SPI SSEL, and the portable code can be cleaned up to take out the manual handling.